### DISCUSSION TITLE
Optimize or refactoring my code plz?)

### DISCUSSION THREAD

**Discussion by projct1** [spatie/laravel-sitemap#502] [Show and tell]
Hi :)
Please see my code. How can I improve it? Am I using the library features correctly?

```php
namespace App\Console\Commands;

use Carbon\Carbon;
use Illuminate\Console\Command;
use Spatie\Sitemap\Sitemap;
use Spatie\Sitemap\SitemapIndex;

class GenerateSitemap extends Command
{
    protected array $data;

    protected $signature = 'sitemap:generate {--prod}';

    protected $description = 'Generate sitemap';

    public function handle(): void
    {
        $range = $this->option('prod') ? range(PROD_CALENDAR_YEAR_MIN, PROD_CALENDAR_YEAR_MAX) : range(YEAR_MIN, YEAR_MAX);

        foreach ($range as $year) {
            $this->data['year'][] = $this->getRoute('year', $year);

            $this->addQuartersAndMonths($year);

            if ($this->option('prod') === false) {
                $this->addWeeks($year);
            }
        }

        $this->save();
    }

    protected function addQuartersAndMonths(int $year): void
    {
        foreach (getStructOfYear() as $quarter => $months) {
            $this->data['quarter'][] = $this->getRoute('quarter', [$year, $quarter]);

            foreach ($months as $month) {
                $this->data['month'][] = routeToMonth($month, $year, $this->option('prod'));
            }
        }
    }

    protected function addWeeks(int $year): void
    {
        foreach (range(1, Carbon::create($year)->isoWeeksInYear) as $week) {
            $this->data['week'][] = $this->getRoute('week', [$year, $week]);
        }
    }

    protected function getRoute(string $route, array | int $params): string
    {
        return route(($this->option('prod') ? 'prod.' : '') . $route, $params);
    }

    protected function save(): void
    {
        $index = SitemapIndex::create();

        foreach ($this->data as $key => $items) {
            Sitemap::create()->add($items)->writeToFile(public_path("sitemap-$key.xml"));
            $index->add("sitemap-$key.xml");
        }

        $index->writeToFile(public_path('sitemap.xml'));
    }
}
```

**Comment by ahmed-esmail-1** [comment]
`<?php

namespace App\Console\Commands;

use Carbon\Carbon;
use Illuminate\Console\Command;
use Spatie\Sitemap\Sitemap;
use Spatie\Sitemap\SitemapIndex;

class GenerateSitemap extends Command
{
    protected array $data;

    protected $signature = 'sitemap:generate {--prod}';
    
    protected $description = 'Generate sitemap';

    public function handle(): void
    {
        $range = $this->option('prod') ? range(config('sitemap.prod_year_min'), config('sitemap.prod_year_max')) : range(config('sitemap.year_min'), config('sitemap.year_max'));

        $sitemapService = app()->make(SitemapService::class);

        foreach ($range as $year) {
            $this->data['year'][] = $this->getRoute('year', $year);
            $this->addQuartersAndMonths($year, $sitemapService);

            if ($this->option('prod') === false) {
                $this->addWeeks($year);
            }
        }

        $this->save();
    }

    protected function addQuartersAndMonths(int $year, SitemapService $sitemapService): void
    {
        $struct = $sitemapService->getYearStructure();

        foreach ($struct as $quarter => $months) {
            $this->data['quarter'][] = $this->getRoute('quarter', [$year, $quarter]);

            foreach ($months as $month) {
                $this->data['month'][] = $sitemapService->routeToMonth($month, $year, $this->option('prod'));
            }
        }
    }

    protected function addWeeks(int $year): void
    {
        foreach (range(1, Carbon::create($year)->isoWeeksInYear) as $week) {
            $this->data['week'][] = $this->getRoute('week', [$year, $week]);
        }
    }

    protected function getRoute(string $route, array | int $params): string
    {
        return route(($this->option('prod') ? 'prod.' : '') . $route, $params);
    }

    protected function save(): void
    {
        $index = SitemapIndex::create();

        foreach ($this->data as $sitemapType => $items) {
            $sitemap = Sitemap::create()->add($items)->writeToFile(public_path("sitemap-$sitemapType.xml"));
            $index->add("sitemap-$sitemapType.xml");

            // Handle errors when writing to file
            if (!$sitemap) {
                $this->error("Failed to write sitemap for $sitemapType");
            }
        }

        $index->writeToFile(public_path('sitemap.xml'));
    }
}
`

**Reply by projct1** [reply]
Edit plz with syntax highlight.

**Reply by ahmed-esmail-1** [reply]
You can ask your question to my Coworker
[https://chat.openai.com/](url)

**Reply by projct1** [reply]
Wrong link 404.

**Reply by ahmed-esmail-1** [reply]
It is working but copy paste it into the browser

**Reply by projct1** [reply]
What copy?
This `url` https://skr.sh/sNM0xlzGdly?
Or chat.openai.com https://skr.sh/sNM5Ii1FZ81?

**Reply by ahmed-esmail-1** [reply]
https://www.openai.com/
